Wpf documentation for 4.8 features and bug fixes (#968)
Microsoft/dotnet: Documentation/compatibility/wpf-accessibility-improvements-48.md

```csharp
    public class MyButtonAutomationPeer : ButtonAutomationPeer 
    { 
        protected override int GetSizeOfSetCore() 
        { 
            // Call into your own logic to provide a value for SizeOfSet 
            return CalculateSizeOfSet(); 
        } 

         protected override int GetPositionInSetCore() 
        { 
            // Call into your own logic to provide a value for PositionInSet 
            return CalculatePositionInSet(); 
        } 
    } 
```
3. __Automatic Values__<br/>Items in [ItemsControls](xref:System.Windows.Controls.ItemsControl) will provide a value for these properties automatically without additional action from the developer. If an ItemsControl is grouped, the collection of groups will be represented as a set and each group counted as a separate set, with each item inside that group providing it�s position inside that group as well as the size of the group. Automatic values are not affected by virtualization. Even if an item is not realized, it is still counted towards the total size of the set and affects the position in the set of it�s sibling items.

> ##### ControllerFor property support

Reviewer: ```suggestion
**ControllerFor property support**
```

> UIAutomation�s ControllerFor property returns an array of automation elements that are manipulated by the automation element that supports this property. This property is commonly used for Auto-suggest accessibility. ControllerFor is used when an automation element affects one or more segments of the application UI or the desktop. Otherwise, it is hard to associate the impact of the control operation with UI elements. This feature adds the ability for controls to provide a value for ControllerFor property.

Reviewer: ```suggestion
UIAutomation’s ControllerFor property returns an array of automation elements that are manipulated by the automation element that supports this property. This property is commonly used for Auto-suggest accessibility. ControllerFor is used when an automation element affects one or more segments of the application UI or the desktop. Otherwise, it is hard to associate the impact of the control operation with UI elements. This feature adds the ability for controls to provide a value for the ControllerFor property.
```

> A new virtual method has been added to [AutomationPeer](xref:System.Windows.Automation.Peers.AutomationPeer):

Reviewer: ```suggestion
A new virtual method has been added to the <xref:System.Windows.Automation.Peers.AutomationPeer> class:
```

```csharp
    virtual protected List<AutomationPeer> GetControlledPeersCore()
```

> To provide a value for the ControllerFor property, simply override this method and return a list of [AutomationPeers](xref:System.Windows.Automation.Peers.AutomationPeer) for the controls being manipulated by this [AutomationPeer](xref:System.Windows.Automation.Peers.AutomationPeer):

Reviewer: ```suggestion
To provide a value for the ControllerFor property, simply override this method and return a list of [AutomationPeers](xref:System.Windows.Automation.Peers.AutomationPeer) for the controls being manipulated by this <xref:System.Windows.Automation.Peers.AutomationPeer>:
```

```csharp

>     

Reviewer: ```suggestion
 ```

public class AutoSuggestTextBox: TextBox
    {
        protected override AutomationPeer OnCreateAutomationPeer()
        {
            return new AutoSuggestTextBoxAutomationPeer(this);
        }

        public ListBox SuggestionListBox;
    }

    internal class AutoSuggestTextBoxAutomationPeer : TextBoxAutomationPeer
    {
        public AutoSuggestTextBoxAutomationPeer(AutoSuggestTextBox owner) : base(owner)
        {        
        }

        protected override List<AutomationPeer> GetControlledPeersCore()
        {
            List<AutomationPeer> controlledPeers = new List<AutomationPeer>();
            AutoSuggestTextBox owner = Owner as AutoSuggestTextBox;

            controlledPeers.Add(UIElementAutomationPeer.CreatePeerForElement(owner.SuggestionListBox));
            return controlledPeers;
        }
    }
```

> ##### Name Automation property is now correctly propagated for DatePicker and RibbonComboBox controls

Reviewer: ```suggestion
**The Name Automation property is now correctly propagated for the DatePicker and RibbonComboBox controls**
```

> Fixed an accessibility problem where [DatePicker](xref:System.Windows.Controls.DatePicker) and [RibbonComboBox](xref:System.Windows.Controls.Ribbon.RibbonComboBox) wouldn't reflect the [AutomationProperties.Name](xref:System.Windows.Automation.AutomationProperties.Name) value assigned to it when read by a screen reader.

Reviewer: ```suggestion
Starting with .NET Framework 4.8, the <xref:System.Windows.Controls.DatePicker> and <xref:System.Windows.Controls.Ribbon.RibbonComboBox> controls reflect the <xref:System.Windows.Automation.AutomationProperties.Name?displayProperty=nameWithType> value assigned to them when read by a screen reader.
```

> ##### Expanders in GroupItem headers are now announced by screen readers

Reviewer: ```suggestion
**Expanders in GroupItem headers are now announced by screen readers**
```

> Fixed an accessibility problem where an [Expander](xref:System.Windows.Controls.Expander) in a re-templated [GroupItem's](xref:System.Windows.Controls.GroupItem) header was not being announced by screen readers.

Reviewer: ```suggestion
Starting with .NET Framework 4.8, an <xref:System.Windows.Controls.Expander> in a re-templated <xref:System.Windows.Controls.GroupItem> header is announced by screen readers.
```

> ##### Improved focus move logic when WPF is hosted in Windows Forms

Reviewer: ```suggestion
**Improved focus move logic when WPF is hosted in Windows Forms**
```

> Fixed an accessibility problem where focus loops inside a WPF [UserControl](xref:System.Windows.Controls.UserControl) instead of breaking out of it under some hosting scenarios

Reviewer: ```suggestion
The focus no longer loops inside a WPF <xref:System.Windows.Controls.UserControl> instead of breaking out of it under some hosting scenarios.
```

> ##### Changed padding header in GridView to not focusable

Reviewer: ```suggestion
**Changed the padding header in GridView to not focusable**
```

> Fixed an accessibility problem where the automatically generated padding header in [GridView](xref:System.Windows.Controls.GridView) was focusable, creating a bad experience for keyboard-only users.

Reviewer: ```suggestion
In the .NET Framework 4.7.2 and earlier versions, the automatically generated padding header in a <xref:System.Windows.Controls.GridView> control was focusable, creating a bad experience for keyboard-only users. Starting with .NET Framework 4.8, it is not focusable.
```

> ##### Fixed automation tree for plain ItemsControl

Reviewer: **Fixed automation tree for plain ItemsControl**
```

> Fixed an accessibility problem where the automation tree exposed for a plain [ItemsControls](xref:System.Windows.Controls.ItemsControl) (as opposed to a derived class like ListBox, DataGrid, etc.) was not correct.

Reviewer: ```suggestion
Fixed an accessibility problem where the automation tree exposed for a plain <xref:System.Windows.Controls.ItemsControl> (as opposed to a derived class like ListBox or DataGrid) was not correct.
```
Can this be described as something other than a fix?

> ##### Fixed bounding rectangles in Per-Monitor DPI.

Reviewer: ```suggestion
**Fixed bounding rectangles in Per-Monitor DPI**
```

> Fixed an accessibility problem where certain bounding rectangles around text are drawn incorrectly by Accessibility tools when WPF is run in Per-Monitor Aware mode.

Reviewer: ```suggestion
In .NET Framework 4.7.2 and earlier versions, certain bounding rectangles around text are drawn incorrectly by Accessibility tools when WPF is run in Per-Monitor Aware mode. Starting with .NET Framework 4.8, they are drawn correctly.
```

</br>

- [x] Quirked
- [ ] Build-time break

### Recommended Action
__How to opt in or out of these changes__
  
In order for the application to benefit from these changes, it must run on the .NET Framework 4.8 or later. The application can benefit from these changes in either of the following ways:

> - It is recompiled to target the .NET Framework 4.8. These accessibility changes are enabled by default on Windows Forms applications that target the .NET Framework 4.8 or later.

Reviewer: ```suggestion
- It is recompiled to target the .NET Framework 4.8. These accessibility changes are enabled by default on WPF applications that target the .NET Framework 4.8 or later.
```

> - It targets the .NET Framework 4.7.2 or earlier version and opts out of the legacy accessibility behaviors by adding the following [AppContext Switch](https://docs.microsoft.com/dotnet/framework/configure-apps/file-schema/runtime/appcontextswitchoverrides-element) to the `<runtime>` section of the app config file and setting it to `false`, as the following example shows.

Reviewer: ```suggestion
- It targets the .NET Framework 4.7.2 or an earlier version and opts out of the legacy accessibility behaviors by adding the following [AppContext switch](https://docs.microsoft.com/dotnet/framework/configure-apps/file-schema/runtime/appcontextswitchoverrides-element) to the `<runtime>` section of the app config file and setting it to `false`, as the following example shows.
```

```xml
    <?xml version="1.0" encoding="utf-8"?>
    <configuration>
      <startup>
        <supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.7"/>
      </startup>
      <runtime>
        <!-- AppContextSwitchOverrides value attribute is in the form of 'key1=true|false;key2=true|false  -->
        <AppContextSwitchOverrides value="Switch.UseLegacyAccessibilityFeatures=false;Switch.UseLegacyAccessibilityFeatures.2=false;Switch.UseLegacyAccessibilityFeatures.3=false" />
      </runtime>
    </configuration>
```    

> Note that to opt in to the accessibility features added in .NET Framework 4.8, you must also opt in to accessibility features of .NET Framework 4.7.2 and .NET Framework 4.7.1 as well.

Reviewer: ```suggestion
Note that to opt in to the accessibility features added in .NET Framework 4.8, you must also opt in to the accessibility features of .NET Framework 4.7.2 and .NET Framework 4.7.1 as well.
```
